core/contracts/financial_templates/implementation/Liquidatable.sol

```diff
@@ -325,10 +343,14 @@ contract Liquidatable is PricelessPositionManager {
             liquidation.settlementPrice
         );
 
+        // TODO: there's a case where the fees cut into the collateral to such a degree where there is not enough to
+        // pay out the TRV in the DisputeSucceeded case. This is unlikely (fees would have to be high). We probably
+        // need to handle this case, however.
         // Calculate rewards as a function of the TRV.
+        FixedPoint.Unsigned memory lockedCollateral = _getCollateral(liquidation.rawLockedCollateral);
         FixedPoint.Unsigned memory disputerDisputeReward = disputerDisputeRewardPct.mul(tokenRedemptionValue);
         FixedPoint.Unsigned memory sponsorDisputeReward = sponsorDisputeRewardPct.mul(tokenRedemptionValue);
-        FixedPoint.Unsigned memory disputeBondAmount = liquidation.lockedCollateral.mul(disputeBondPct);
+        FixedPoint.Unsigned memory disputeBondAmount = lockedCollateral.mul(disputeBondPct);
```

Comment:
As discussed IRL, it is possible that the regular fee is so large that during the liveness of the liquidation, the fee accrues to a large enough value that rewards can't be paid out. This is relevant in the case where a dispute succeeded.

In other words, we need to enforce the variant that `(fees + sponsor-reward + disputer-reward) <= TRV`. 

Absent regular fees, the rewards are taken as a % of TRV, however if this invariant is violated, then there will not be enough collateral remaining in the contract to repay all of the parties their owed amount:
- Disputer should get `disputer reward (% of TRV) + dispute bond`
- Sponsor should get `(locked-collateral - TRV) + sponsor reward (% of TRV)`
- Liquidator should get `TRV - disputer-reward - sponsor-reward`

One way around this could be to multiply `tokenRedemptionValue` by `cumulativeFeeMultiplier` so that the rewards are taken as a % of the (TRV - fees). 
- Disputer should get `disputer reward (% of TRV - fees) + dispute bond`
- Sponsor should get `(locked-collateral - (TRV - fees)) + sponsor reward (% of TRV - fees)`
- Liquidator should get `TRV - fees - disputer-reward - sponsor-reward`

This however has the edge case where `fees > TRV`.  We could instead decide to take rewards out of the `lockedCollateral` but this would require us check new edge cases where `rewards > (collateral - TRV)` (i.e. rewards are greater than excess collateral owed to sponsor, EVEN if total reward % <= 1)

Perhaps we could have `FeePayer` refuse to pay fees once `cumulativeFeeMultiplier < (TRV / lockedCollateral)`

Reply:
> One way around this could be to multiply `tokenRedemptionValue` by `cumulativeFeeMultiplier` so that the rewards are taken as a % of the (TRV - fees).

I think it would be slightly more complicated than that. The fees are cumulative, so you'd have to snapshot the cumulativeFeeMultiplier when the liquidation begins so you could divide by the initial value and then multiply by the later value. Otherwise you'd be paying the cumulative fees over the life of the contract.

Also what are you defining as fees here? Are fees the fees that are charged on just the TRV portion of the collateral or the entire locked collateral (TRV + extra sponsor collateral)?

Generally the way I think of fees is as a % of collateral. So if you charge a 5% fee on the entire contract between the beginning of liquidation and the withdrawal, then every subdivision of collateral should become 5% smaller. In that case, a 5% fee would just make the TRV 5% smaller, but it would also make the excess sponsor collateral 5% smaller. So if you just scaled all the payouts down by 5%, it would all work out. And since the fees cannot be >100%, the worst case is a 100% fee that sets the payouts to 0.

> 
> This however has the edge case where `fees > TRV`. We could instead decide to take rewards out of the `lockedCollateral` but this would require us check new edge cases where `rewards > (collateral - TRV)` (i.e. rewards are greater than excess collateral owed to sponsor, EVEN if total reward % <= 1)

If you're trying to have the TRV pay for the fees on all the locked collateral that case exists. However, if the TRV is only paying fees on the TRV portion, then a > 100% fee is impossible IIUC.

Reply:
@nicholaspai, I think this discussion is great, just want to make sure we're on the same page that fixing this edge case should go in a follow up PR.

Reply:
I'm defining fees as just all regular fees paid out, which are deducted from the `collateral.rawValue`. I think we have the same definition of fees.

Scaling everything down by fees (as % of collateral) would also work--this isn't in the code yet though right?

Reply:
> @nicholaspai, I think this discussion is great, just want to make sure we're on the same page that fixing this edge case should go in a follow up PR?

Do you think it should be in a follow up PR because of the sheer amount of code changes that would address this edge case? because functionally I would argue that it is a known edge case within the scope of this PR. But practically if its a lot of code then I think it would be wise to separate the code so that errors don't get lost in the mass changes.

Reply:
> Scaling everything down by fees (as % of collateral) would also work--this isn't in the code yet though right?

Correct, that isn't implemented. We should talk about it more offline, but after thinking about it a little more, scaling everything by the fee percentage seems to be the simplest to understand and have the fewest edge cases. My only worry would be that it may be complicated to implement.

> Do you think it should be in a follow up PR because of the sheer amount of code changes that would address this edge case? because functionally I would argue that it is a known edge case within the scope of this PR. But practically if its a lot of code then I think it would be wise to separate the code so that errors don't get lost in the mass changes.

I think these changes may be subtle and I don't want them getting lost in the hubbub of a larger PR.

Reply:
ok that's reasonable. Why do you think "scaling everything by the fee percentage" would be complicated to implement?

Reply:
I can't think of a way to do it without snapshotting an additional variable at liquidation time.